File: chapter3/recursion.rb
# even?メソッドがもうあるのでis_even?メソッドとする
def is_even?(n)
  if n == 0
    true
  elsif n == 1
    false
  elsif n > 1
    is_even?(n-2)
  elsif n < 0
    is_even?(-n)
  end
end

# そもそも2進数の論理積を取れば再帰は不要
def is_even2?(n)
  n & 1 == 0
end

File: chapter3/recursion.js
{
  // 負の数未対応
  const isEven = n => {
    if (n === 0) {
      return true
    } else if( n === 1) {
      return false
    } else if ( n > 1) {
      return isEven(n-2)
    }
  }
}
// 負の数にも対応
{
  const isEven = n => {
    if (n === 0) return true
    else if (n === 1) return false
    else if (n < 0) return isEven(-n)
    else if (n > 1) return isEven(n)
  }
}

// 再帰ではなく2進数を使う
const isEven = n => (n & 1) == 0

